### PR TITLE
feat(config/validation): validate options which support regex/glob matching

### LINKED ISSUE #28554: Throw config validation error if combining * with other values in a matchRegexOrGlobList field
### Describe the proposed change(s).

Using `matchRepositories` as an example, we should throw a config validation error if the user configures `*` or `**` patterns with any other value.

e.g. these are all not ok:

```
  "matchRepositories": ["*", "!abc/def"]
  "matchRepositories": ["*", "x/y", "!abc/def"]
  "matchRepositories": ["*", "x/y"]
  "matchRepositories": ["*", "**"]
```

We should keep a list of config options/fields which support matchRegexOrGlob - it will increase over time. I think so far it's `allowedHeaders`, `allowedEnvVars`, `autodiscoverProjects` and `matchRepositories`.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add new fields `patternMatch` to the config options, which will be used to identify options which support regex/glob matching
- Refactor validation logic
- Add validation for options which support regex/Glob matching. This validation logic ensures that the options do not contain `*` or `**` values along with other values.

Example: (these are not ok)
```
 "matchRepositories": ["*", "!abc/def"]
  "matchRepositories": ["*", "x/y", "!abc/def"]
  "matchRepositories": ["*", "x/y"]
  "matchRepositories": ["*", "**"]
```

## Context
Closes: https://github.com/renovatebot/renovate/issues/28554

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/types.ts [hunk 1]
```diff
@@ -442,6 +442,8 @@ export interface RenovateOptionBase {
    * This is used to add depreciation message in the docs
    */
   deprecationMsg?: string;
+
+  regexOrGlob?: boolean;
```
Comment (secustor) [inline lib/config/types.ts:446]:
```suggestion
  patternMatch?: boolean;
```
This is more generic, when we evolve this in the future.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
This new field needs to be explained in the docs. Right now it only shows the field in the config options table.

**Comment by RahulGautamSingh** [comment]
> This new field needs to be explained in the docs. Right now it only shows the field in the config options table.

What place should this be added to?

File: lib/config/types.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/types.ts:446]: Not sure about this, aren't regex patterns too.

Comment (secustor) [reply lib/config/types.ts:446]: The docs page is called String Pattern Matching and contains regex and glob, so I think this should be clear. 
https://docs.renovatebot.com/string-pattern-matching/

**Comment by secustor** [comment]
> This new field needs to be explained in the docs. Right now it only shows the field in the config options table.
This is not exposed to users, but rather for internal use to identify options which support the pattern matching feature. 

Later this can be used to automatically link from these user exposed options to our "pattern matching" docs

**Comment by RahulGautamSingh** [comment]
> This is not exposed to users, but rather for internal use to identify options which support the pattern matching feature.
> 
> Later this can be used to automatically link from these user exposed options to our "pattern matching" docs

Should I ignore adding it to the docs for now? ie. not show it in the config options table

**Comment by RahulGautamSingh** [comment]
The docs part needs to have some discussion so I will keep this option only for internal use in the context of this PR.

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs validation added

File: lib/config/types.ts [hunk 2]
```diff
@@ -442,6 +442,11 @@ export interface RenovateOptionBase {
    * This is used to add depreciation message in the docs
    */
   deprecationMsg?: string;
+
+  /**
+   * For internal use only: add it any option that supports regex or glob matching
```
Comment (HonkingGoose) [inline lib/config/types.ts:447]:
```suggestion
   * For internal use only: add it to any config option that supports regex or glob matching
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/validation-helpers/regex-glob-matchers.spec.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    const res = check({
+      val: ['*', '**'],
+      currentPath: 'hostRules[0].allowedHeaders',
+    });
+    expect(res).toHaveLength(1);
+  });
+
+  it('should have errors - 2', () => {
+    const res = check({
+      val: ['*', 2] as never,
```
Comment (viceice) [inline lib/config/validation-helpers/regex-glob-matchers.spec.ts:14]: use ts-expect-error instead

File: lib/config/validation.ts [hunk 4]
```diff
@@ -111,6 +113,18 @@ function isInhertConfigOption(key: string): boolean {
   return optionInherits.has(key);
 }
 
+function isRegexOrGlobOption(key: string): boolean {
+  if (!optionRegexOrGlob) {
+    optionRegexOrGlob = new Set();
+    for (const option of options) {
```
Comment (viceice) [inline lib/config/validation.ts:119]: we should extract a single function to init, so we don't need to iterate multiple times

File: lib/config/validation-helpers/regex-glob-matchers.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import is from '@sindresorhus/is';
+import type { ValidationMessage } from '../types';
+import type { CheckMatcherArgs } from './types';
+
+/**
+ * Only if type condition or context condition violated then errors array will be mutated to store metadata
+ */
+export function check({
+  val,
```
Comment (viceice) [inline lib/config/validation-helpers/regex-glob-matchers.ts:9]: why not except unknown here, makes type checks simpler

File: lib/config/validation.ts (hunk 4 above)
Comment (rarkins) [reply lib/config/validation.ts:119]: Yes, move the set outside this function if the function is called multiple times

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/validation-helpers/types.ts [hunk 6]
```diff
@@ -4,3 +4,8 @@ export interface CheckManagerArgs {
   resolvedRule: PackageRule;
   currentPath: string;
 }
+
+export interface CheckMatcherArgs {
+  val: unknown[];
```
Comment (viceice) [inline lib/config/validation-helpers/types.ts:9]:
```suggestion
  val: unknown;
```
should work too, `is.array` will narrow the type

File: tools/docs/config.ts [hunk 7]
```diff
@@ -94,6 +94,7 @@ function genTable(obj: [string, string][], type: string, def: any): string {
     'experimentalIssues',
     'advancedUse',
     'deprecationMsg',
+    'patternMatch',
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply tools/docs/config.ts:97]: No, adding it to the `ignoredKeys` prevents the field from being added to the table.

**Comment by RahulGautamSingh** [comment]
Should the PR title be renamed to something like `feat(validation): validate regexOrGlob options`? Since, the main objective of this PR is to validate the regex/glob matching options and, the new field `patternMatch` is being added to facilitate it.

Another PR which will add the description of the new field to the docs and also add the field to the config options table can be named as `feat: patternMatch` I think.

File: lib/config/validation.spec.ts [hunk 8]
```diff
@@ -1225,6 +1225,30 @@ describe('config/validation', () => {
       expect(warnings).toHaveLength(0);
       expect(errors).toHaveLength(2);
     });
+
+    it('catches when * or ** is combined with others patterns in a regexOrGlob option', async () => {
+      const config = {
+        packageRules: [
+          {
+            matchRepositories: ['*', 'repo'],
```
Comment (secustor) [inline lib/config/validation.spec.ts:1233]:
```suggestion
            matchRepositories: ['groupA/**', 'groupB/**'],
```
We should add a positive check that we allow multiple of such patterns.

File: lib/config/validation.spec.ts [hunk 9]
```diff
@@ -1225,6 +1225,30 @@ describe('config/validation', () => {
...
+        config,
+      );
+      expect(errors).toMatchObject([
+        {
+          message:
+            'packageRules[0].matchRepositories: Your input contains * or ** along with other patterns. Please remove them, as * or ** matches all patterns.',
+          topic: 'Configuration Error',
+        },
+      ]);
+      expect(errors).toHaveLength(1);
```
Comment (secustor) [inline lib/config/validation.spec.ts:1249]: This shouldn't throw errors as far I know.

**Review by secustor — DISMISSED** [review]

File: lib/config/validation.ts [hunk 10]
```diff
@@ -100,27 +103,45 @@ function getDeprecationMessage(option: string): string | undefined {
...
-    optionGlobals = new Set();
-    for (const option of options) {
-      if (option.globalOnly) {
-        optionGlobals.add(option.name);
-      }
+  return optionGlobals.has(key);
+}
+
+function initOptions(): void {
+  optionParents = {};
```
Comment (viceice) [inline lib/config/validation.ts:118]:
```suggestion
  if (optionsInitialized) {
    return;
  }
  optionParents = {};
```

what about to add the check here?

File: lib/config/validation.ts [hunk 11]
```diff
@@ -139,20 +160,10 @@ export async function validateConfig(
...
-    optionParents = {};
-    options.forEach((option) => {
-      if (option.parents) {
-        optionParents[option.name] = option.parents;
-      }
-    });
+  if (!optionsInitialized) {
+    initOptions();
   }
+
```
Comment (viceice) [inline lib/config/validation.ts:166]:
```suggestion
  initOptions();

```

**Review by viceice — COMMENTED** [review]
otherwise lgtm

**Review by viceice — DISMISSED** [review]

**Review by secustor — DISMISSED** [review]

File: tools/docs/config.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply tools/docs/config.ts:97]: Let's do that later. We also need to standardize our terminology

**Review by rarkins — COMMENTED** [review]
Can we combine this with other checks that validate the regex is ok?

**Comment by RahulGautamSingh** [comment]
> Can we combine this with other checks that validate the regex is ok?

We have two checks for regex.
1. `isRegexMatch` combined with `getRegexPredicate`, which is used to validation the regex for options of `type:string`
2. `matchRegexOrGlobList` which is used to check whether the`env` key names & `hostRules.headers` set in repo config match the patterns present in `allowedEnv` & `allowedHeaders` from the admin config resp.

For check(1), there isn't any way to combine as it only checks for regex validation only, no glob matching and it is designed for single string rather than an array

For check(2), we could make it work if we modify fn to include an if-check for the * or ** logic and expand the return types, to include an extra field say `errMsg` which we can use to know about the error.

For eg.
Currently the `matchRegexOrGlobList` fn returns a boolean
https://github.com/renovatebot/renovate/blob/ca23445c274fa6802f12ce564e8f7919a0d1a0f3/lib/util/string-match.ts#L26-L29
`true`:  if the key is matches the patterns
`false`: if no match is found or patterns list is empty

New return types: 
```ts
export function matchRegexOrGlobList(
  input: string,
  patterns: string[],
  currentPath?: string,
): { res: boolean; errMsg?: string } 
```
Incase we encounter an error say `no matc found` or `* or ** string combined with other patterns`, we pass it using the `errMsg` field and it can be then used in the validation logic.

**Comment by rarkins** [comment]
@RahulGautamSingh I pushed my suggested change to this branch - wdyt?

**Comment by RahulGautamSingh** [comment]
> @RahulGautamSingh I pushed my suggested change to this branch - wdyt?

Looks good.

File: lib/config/validation-helpers/regex-glob-matchers.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      res.push({
+        topic: 'Configuration Error',
+        message: `${currentPath}: Your input contains * or ** along with other patterns. Please remove them, as * or ** matches all patterns.`,
+      });
+    }
+    for (const matcher of matchers) {
+      // Validate regex pattern
+      if (isRegexMatch(matcher)) {
+        const autodiscoveryPred = getRegexPredicate(matcher);
+        if (!autodiscoveryPred) {
```
Comment (RahulGautamSingh) [inline lib/config/validation-helpers/regex-glob-matchers.ts:29]:
```suggestion
        if (!getRegexPredicate(matcher)) {
```

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**